### PR TITLE
CHANGE (GraphEngine): @W-13123571@: Handle method invocations made directly on iterated array item

### PR DESCRIPTION
**Pull request by rmohan20**
Previously, this worked as expected:
```
Bean b1 = beans[i];
System.debug(b1.getValue());
```

But this wouldn't:
```
System.debug(beans[i].getValue());
```

Changes ensure that paths are expanded when a method call is made directly on an array item. Also includes tests to cover various aspects of loops and method calls on iterated items.

### CODE REVIEW

File: sfge/src/main/java/com/salesforce/graph/ops/MethodUtil.java [hunk 1]
```diff
@@ -635,6 +641,7 @@ public static List<ApexPath> getPaths(
                 final String methodName = methodCallExpression.getMethodName();
                 String fullMethodName = methodCallExpression.getFullMethodName();
                 if (methodName.equals(fullMethodName)) {
+                    // TODO: Method invoked on ArrayLoadExpression is incorrectly getting in here
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/MethodUtil.java:644]: Will cleanup this todo. Doesn't matter anymore.

File: sfge/src/main/java/com/salesforce/graph/ops/MethodUtil.java [hunk 2]
```diff
@@ -531,6 +527,16 @@ public static Optional<ApexValue<?>> getApexValue(
             apexValue = symbols.getReturnedValue((InvocableVertex) vertex).orElse(null);
         }
 
+        if (apexValue == null && vertex instanceof MethodCallExpressionVertex) {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/MethodUtil.java:530]: This portion makes sure that a method call on a class instance array item gets a corresponding method definition. (See javadoc on test for full explanation)

File: sfge/src/main/java/com/salesforce/graph/symbols/PathScopeVisitor.java [hunk 3]
```diff
@@ -1435,6 +1436,10 @@ private Optional<ApexValue<?>> handleApexValueMethod(MethodCallExpressionVertex
             // The method is being called on a chain
             // SObjectType sot = Schema.getGlobalDescribe().get('Account');
             apexValue = getReturnedValue(previous).orElse(null);
+        } else if (arrayInvocation != null) {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/PathScopeVisitor.java:1439]: This portion makes sure that a method call on a primitive or standard library instance array item gets the corresponding value returned.

File: sfge/src/test/java/com/salesforce/graph/ops/expander/ApexPathExpanderInLoopTest.java [hunk 4]
```diff
@@ -4,18 +4,50 @@
...
 import java.util.List;
 import java.util.stream.Collectors;
 import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
 import org.hamcrest.MatcherAssert;
 import org.hamcrest.Matchers;
 import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 
+/**
```
Comment (rmohan20) [inline sfge/src/test/java/com/salesforce/graph/ops/expander/ApexPathExpanderInLoopTest.java:23]: Javadoc explains the differences between method calls on various types of array items.

File: sfge/src/main/java/com/salesforce/graph/ops/MethodUtil.java [hunk 5]
```diff
@@ -613,7 +619,6 @@ public static List<ApexPath> getPaths(
                                 .orElse(null);
             } else if (vertex instanceof ArrayLoadExpressionVertex) {
                 // Intentionally left blank
-                // TODO: Why?
```
Comment (jfeingold35) [inline sfge/src/main/java/com/salesforce/graph/ops/MethodUtil.java:616]: Can we remove this `else if` block, or does keeping it somehow preserve clarity?

**Review by jfeingold35 — APPROVED** [review]
Approved, but if possible could you please add a JDoc header to each test explaining the specific case it covers?

**Pull request merged**